File: implementations/styx-js/src/lexer.ts
import { Span, ParseError } from "./types.js";

export type TokenType =
  | "scalar"
  | "quoted"
  | "raw"
  | "heredoc"
  | "lbrace"
  | "rbrace"
  | "lparen"
  | "rparen"
  | "comma"
  | "at"
  | "tag"
  | "gt"
  | "newline"
  | "eof";

export interface Token {
  type: TokenType;
  text: string;
  span: Span;
  /** True if there was whitespace before this token */
  hadWhitespaceBefore: boolean;
  /** True if there was a newline before this token */
  hadNewlineBefore: boolean;
}

const SPECIAL_CHARS = new Set(["{", "}", "(", ")", ",", '"', ">", " ", "\t", "\n", "\r"]);

export class Lexer {
  private pos = 0; // character position
  private bytePos = 0; // byte position for spans
  private line = 1;
  private col = 1;

  constructor(private source: string) {}

  private peek(offset = 0): string {
    return this.source[this.pos + offset] ?? "";
  }

  private advance(): string {
    if (this.pos >= this.source.length) return "";

    // Check for surrogate pair (emoji, etc.)
    const code = this.source.charCodeAt(this.pos);
    let ch: string;
    if (code >= 0xd800 && code <= 0xdbff && this.pos + 1 < this.source.length) {
      // High surrogate - consume both code units
      ch = this.source.slice(this.pos, this.pos + 2);
      this.pos += 2;
    } else {
      ch = this.source[this.pos++];
    }

    // Calculate UTF-8 byte length of the character
    this.bytePos += this.utf8ByteLength(ch);
    if (ch === "\n") {
      this.line++;
      this.col = 1;
    } else {
      this.col++;
    }
    return ch;
  }

  private utf8ByteLength(ch: string): number {
    // Use TextEncoder to get accurate UTF-8 byte length
    // This handles surrogate pairs and all Unicode correctly
    return new TextEncoder().encode(ch).length;
  }

  /** Get current byte position for span start */
  private get byteStart(): number {
    return this.bytePos;
  }

  private skipWhitespaceAndComments(): { hadWhitespace: boolean; hadNewline: boolean } {
    let hadWhitespace = false;
    let hadNewline = false;
    while (this.pos < this.source.length) {
      const ch = this.peek();
      if (ch === " " || ch === "\t" || ch === "\r") {
        hadWhitespace = true;
        this.advance();
      } else if (ch === "\n") {
        hadWhitespace = true;
        hadNewline = true;
        this.advance();
      } else if (ch === "/" && this.peek(1) === "/") {
        // Line comment - skip to end of line
        hadWhitespace = true;
        while (this.pos < this.source.length && this.peek() !== "\n") {
          this.advance();
        }
      } else {
        break;
      }
    }
    return { hadWhitespace, hadNewline };
  }

  nextToken(): Token {
    const { hadWhitespace, hadNewline } = this.skipWhitespaceAndComments();

    if (this.pos >= this.source.length) {
      return {
        type: "eof",
        text: "",
        span: { start: this.bytePos, end: this.bytePos },
        hadWhitespaceBefore: hadWhitespace,
        hadNewlineBefore: hadNewline,
      };
    }

    const start = this.bytePos;
    const ch = this.peek();

    // Single-character tokens
    if (ch === "{") {
      this.advance();
      return {
        type: "lbrace",
        text: "{",
        span: { start, end: this.bytePos },
        hadWhitespaceBefore: hadWhitespace,
        hadNewlineBefore: hadNewline,
      };
    }
    if (ch === "}") {
      this.advance();
      return {
        type: "rbrace",
        text: "}",
        span: { start, end: this.bytePos },
        hadWhitespaceBefore: hadWhitespace,
        hadNewlineBefore: hadNewline,
      };
    }
    if (ch === "(") {
      this.advance();
      return {
        type: "lparen",
        text: "(",
        span: { start, end: this.bytePos },
        hadWhitespaceBefore: hadWhitespace,
        hadNewlineBefore: hadNewline,
      };
    }
    if (ch === ")") {
      this.advance();
      return {
        type: "rparen",
        text: ")",
        span: { start, end: this.bytePos },
        hadWhitespaceBefore: hadWhitespace,
        hadNewlineBefore: hadNewline,
      };
    }
    if (ch === ",") {
      this.advance();
      return {
        type: "comma",
        text: ",",
        span: { start, end: this.bytePos },
        hadWhitespaceBefore: hadWhitespace,
        hadNewlineBefore: hadNewline,
      };
    }
    if (ch === ">") {
      this.advance();
      return {
        type: "gt",
        text: ">",
        span: { start, end: this.bytePos },
        hadWhitespaceBefore: hadWhitespace,
        hadNewlineBefore: hadNewline,
      };
    }

    // @ - either unit or tag
    if (ch === "@") {
      this.advance();
      // Check if it's a tag name
      if (this.isTagStart(this.peek())) {
        const nameStart = this.pos;
        while (this.isTagChar(this.peek())) {
          this.advance();
        }
        const name = this.source.slice(nameStart, this.pos);
        return {
          type: "tag",
          text: name,
          span: { start, end: this.bytePos },
          hadWhitespaceBefore: hadWhitespace,
          hadNewlineBefore: hadNewline,
        };
      }
      return {
        type: "at",
        text: "@",
        span: { start, end: this.bytePos },
        hadWhitespaceBefore: hadWhitespace,
        hadNewlineBefore: hadNewline,
      };
    }

    // Quoted string
    if (ch === '"') {
      return this.readQuotedString(start, hadWhitespace, hadNewline);
    }

    // Raw string
    if (ch === "r" && (this.peek(1) === '"' || this.peek(1) === "#")) {
      return this.readRawString(start, hadWhitespace, hadNewline);
    }

    // Heredoc - only if << is followed by uppercase letter
    if (ch === "<" && this.peek(1) === "<") {
      const afterLtLt = this.peek(2);
      if (afterLtLt >= "A" && afterLtLt <= "Z") {
        return this.readHeredoc(start, hadWhitespace, hadNewline);
      }
      // << not followed by uppercase - return error at just <<
      this.advance(); // <
      this.advance(); // <
      const errorEnd = this.bytePos;
      // Skip rest of line for recovery
      while (this.pos < this.source.length && this.peek() !== "\n") {
        this.advance();
      }
      throw new ParseError("unexpected token", { start, end: errorEnd });
    }

    // Bare scalar
    return this.readBareScalar(start, hadWhitespace, hadNewline);
  }

  private isTagStart(ch: string): boolean {
    return /[A-Za-z_]/.test(ch);
  }

  private isTagChar(ch: string): boolean {
    return /[A-Za-z0-9_\-]/.test(ch);
  }

  private readQuotedString(start: number, hadWhitespace: boolean, hadNewline: boolean): Token {
    this.advance(); // opening "
    let text = "";

    while (this.pos < this.source.length) {
      const ch = this.peek();
      if (ch === '"') {
        this.advance();
        return {
          type: "quoted",
          text,
          span: { start, end: this.bytePos },
          hadWhitespaceBefore: hadWhitespace,
          hadNewlineBefore: hadNewline,
        };
      }
      if (ch === "\\") {
        const escapeStart = this.bytePos;
        this.advance();
        const escaped = this.advance();
        switch (escaped) {
          case "n":
            text += "\n";
            break;
          case "r":
            text += "\r";
            break;
          case "t":
            text += "\t";
            break;

          case "\\":
            text += "\\";
            break;
          case '"':
            text += '"';
            break;
          case "u":
            text += this.readUnicodeEscape();
            break;
          default:
            // Unknown escape - this is a parse error
            throw new ParseError(`invalid escape sequence: \\${escaped}`, {
              start: escapeStart,
              end: this.bytePos,
            });
        }
      } else if (ch === "\n" || ch === "\r") {
        // Unterminated string - include the newline in the span
        this.advance();
        if (ch === "\r" && this.peek() === "\n") {
          this.advance();
        }
        throw new ParseError("unexpected token", { start, end: this.bytePos });
      } else {
        text += this.advance();
      }
    }

    // EOF without closing quote - error
    throw new ParseError("unexpected token", { start, end: this.bytePos });
  }

  private readUnicodeEscape(): string {
    if (this.peek() === "{") {
      this.advance(); // {
      let hex = "";
      while (this.peek() !== "}" && this.pos < this.source.length) {
        hex += this.advance();
      }
      this.advance(); // }
      const codePoint = parseInt(hex, 16);
      return String.fromCodePoint(codePoint);
    } else {
      // \uXXXX format
      let hex = "";
      for (let i = 0; i < 4; i++) {
        hex += this.advance();
      }
      const codePoint = parseInt(hex, 16);
      return String.fromCodePoint(codePoint);
    }
  }

  private readRawString(start: number, hadWhitespace: boolean, hadNewline: boolean): Token {
    this.advance(); // r
    let hashes = 0;
    while (this.peek() === "#") {
      this.advance();
      hashes++;
    }
    this.advance(); // opening "

    let text = "";
    const closePattern = '"' + "#".repeat(hashes);

    while (this.pos < this.source.length) {
      if (this.source.slice(this.pos, this.pos + closePattern.length) === closePattern) {
        // Advance through the close pattern
        for (let i = 0; i < closePattern.length; i++) {
          this.advance();
        }
        return {
          type: "raw",
          text,
          span: { start, end: this.bytePos },
          hadWhitespaceBefore: hadWhitespace,
          hadNewlineBefore: hadNewline,
        };
      }
      text += this.advance();
    }

    throw new ParseError("unclosed raw string", { start, end: this.bytePos });
  }

  private readHeredoc(start: number, hadWhitespace: boolean, hadNewline: boolean): Token {
    this.advance(); // <
    this.advance(); // <

    // Read delimiter (and optional language hint)
    let delimiter = "";
    while (this.pos < this.source.length && this.peek() !== "\n") {
      delimiter += this.advance();
    }
    if (this.pos < this.source.length) {
      this.advance(); // newline
    }

    // Track content start (after the opening line)
    const contentStart = this.bytePos;

    // Read content until delimiter on its own line
    let text = "";
    const bareDelimiter = delimiter.split(",")[0];

    while (this.pos < this.source.length) {
      // Read a line
      let line = "";
      while (this.pos < this.source.length && this.peek() !== "\n") {
        line += this.advance();
      }

      // Check for exact match (no indentation)
      if (line === bareDelimiter) {
        // Language hint is metadata only, does not affect content (r[scalar.heredoc.lang])
        return {
          type: "heredoc",
          text,
          span: { start, end: this.bytePos },
          hadWhitespaceBefore: hadWhitespace,
          hadNewlineBefore: hadNewline,
        };
      }

      // Check for indented closing delimiter
      const stripped = line.replace(/^[ \t]+/, "");
      if (stripped === bareDelimiter) {
        const indentLen = line.length - stripped.length;
        // Dedent the content by stripping up to indentLen from each line
        const result = this.dedentHeredoc(text, indentLen);
        // Language hint is metadata only, does not affect content (r[scalar.heredoc.lang])
        return {
          type: "heredoc",
          text: result,
          span: { start, end: this.bytePos },
          hadWhitespaceBefore: hadWhitespace,
          hadNewlineBefore: hadNewline,
        };
      }

      // Add line to content
      text += line;
      if (this.pos < this.source.length && this.peek() === "\n") {
        this.advance();
        text += "\n";
      }
    }

    // Heredoc without closing delimiter - error points at the unmatched content
    throw new ParseError("unexpected token", { start: contentStart, end: this.bytePos });
  }

  /** Strip up to indentLen whitespace characters from the start of each line. */
  private dedentHeredoc(content: string, indentLen: number): string {
    const lines = content.split("\n");
    const result: string[] = [];
    for (const line of lines) {
      let stripped = 0;
      for (const ch of line) {
        if (stripped >= indentLen) {
          break;
        }
        if (ch === " " || ch === "\t") {
          stripped++;
        } else {
          break;
        }
      }
      result.push(line.slice(stripped));
    }
    return result.join("\n");
  }

  private readBareScalar(start: number, hadWhitespace: boolean, hadNewline: boolean): Token {
    let text = "";
    while (this.pos < this.source.length) {
      const ch = this.peek();
      if (SPECIAL_CHARS.has(ch)) {
        break;
      }
      text += this.advance();
    }
    return {
      type: "scalar",
      text,
      span: { start, end: this.bytePos },
      hadWhitespaceBefore: hadWhitespace,
      hadNewlineBefore: hadNewline,
    };
  }
}

File: implementations/styx-py/styx/lexer.py
"""Lexer for Styx configuration language."""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import Enum
from typing import Final

from .types import ParseError, Span


class TokenType(Enum):
    """Token types."""

    SCALAR = "scalar"
    QUOTED = "quoted"
    RAW = "raw"
    HEREDOC = "heredoc"
    LBRACE = "lbrace"
    RBRACE = "rbrace"
    LPAREN = "lparen"
    RPAREN = "rparen"
    COMMA = "comma"
    AT = "at"
    TAG = "tag"
    GT = "gt"
    EOF = "eof"


@dataclass(slots=True)
class Token:
    """A lexer token."""

    type: TokenType
    text: str
    span: Span
    had_whitespace_before: bool
    had_newline_before: bool


SPECIAL_CHARS: Final[frozenset[str]] = frozenset(
    ["{", "}", "(", ")", ",", '"', ">", " ", "\t", "\n", "\r"]
)

TAG_START_RE: Final[re.Pattern[str]] = re.compile(r"[A-Za-z_]")
TAG_CHAR_RE: Final[re.Pattern[str]] = re.compile(r"[A-Za-z0-9_\-]")


class Lexer:
    """Tokenizer for Styx source code."""

    __slots__ = ("byte_pos", "pos", "source")

    def __init__(self, source: str) -> None:
        self.source = source
        self.pos = 0  # character position
        self.byte_pos = 0  # byte position for spans

    def _peek(self, offset: int = 0) -> str:
        """Look ahead in the source."""
        idx = self.pos + offset
        if idx >= len(self.source):
            return ""
        return self.source[idx]

    def _advance(self) -> str:
        """Consume and return the next character."""
        if self.pos >= len(self.source):
            return ""
        ch = self.source[self.pos]
        self.pos += 1
        self.byte_pos += len(ch.encode("utf-8"))
        return ch

    def _skip_whitespace_and_comments(self) -> tuple[bool, bool]:
        """Skip whitespace and comments, return (had_whitespace, had_newline)."""
        had_whitespace = False
        had_newline = False

        while self.pos < len(self.source):
            ch = self._peek()
            if ch in (" ", "\t", "\r"):
                had_whitespace = True
                self._advance()
            elif ch == "\n":
                had_whitespace = True
                had_newline = True
                self._advance()
            elif ch == "/" and self._peek(1) == "/":
                had_whitespace = True
                while self.pos < len(self.source) and self._peek() != "\n":
                    self._advance()
            else:
                break

        return had_whitespace, had_newline

    def _is_tag_start(self, ch: str) -> bool:
        """Check if character can start a tag name."""
        return bool(TAG_START_RE.match(ch))

    def _is_tag_char(self, ch: str) -> bool:
        """Check if character can be in a tag name."""
        return bool(TAG_CHAR_RE.match(ch))

    def next_token(self) -> Token:
        """Return the next token."""
        had_whitespace, had_newline = self._skip_whitespace_and_comments()

        if self.pos >= len(self.source):
            return Token(
                type=TokenType.EOF,
                text="",
                span=Span(self.byte_pos, self.byte_pos),
                had_whitespace_before=had_whitespace,
                had_newline_before=had_newline,
            )

        start = self.byte_pos
        ch = self._peek()

        # Single-character tokens
        match ch:
            case "{":
                self._advance()
                return Token(
                    TokenType.LBRACE, "{", Span(start, self.byte_pos), had_whitespace, had_newline
                )
            case "}":
                self._advance()
                return Token(
                    TokenType.RBRACE, "}", Span(start, self.byte_pos), had_whitespace, had_newline
                )
            case "(":
                self._advance()
                return Token(
                    TokenType.LPAREN, "(", Span(start, self.byte_pos), had_whitespace, had_newline
                )
            case ")":
                self._advance()
                return Token(
                    TokenType.RPAREN, ")", Span(start, self.byte_pos), had_whitespace, had_newline
                )
            case ",":
                self._advance()
                return Token(
                    TokenType.COMMA, ",", Span(start, self.byte_pos), had_whitespace, had_newline
                )
            case ">":
                self._advance()
                return Token(
                    TokenType.GT, ">", Span(start, self.byte_pos), had_whitespace, had_newline
                )

        # @ - either unit or tag
        if ch == "@":
            self._advance()
            if self._is_tag_start(self._peek()):
                name_start = self.pos
                while self._is_tag_char(self._peek()):
                    self._advance()
                name = self.source[name_start : self.pos]
                return Token(
                    TokenType.TAG, name, Span(start, self.byte_pos), had_whitespace, had_newline
                )
            return Token(TokenType.AT, "@", Span(start, self.byte_pos), had_whitespace, had_newline)

        # Quoted string
        if ch == '"':
            return self._read_quoted_string(start, had_whitespace, had_newline)

        # Raw string
        if ch == "r" and self._peek(1) in ('"', "#"):
            return self._read_raw_string(start, had_whitespace, had_newline)

        # Heredoc - only if << is followed by uppercase letter
        if ch == "<" and self._peek(1) == "<":
            after_lt_lt = self._peek(2)
            if after_lt_lt.isupper():
                return self._read_heredoc(start, had_whitespace, had_newline)
            # << not followed by uppercase - return error at just <<
            self._advance()  # <
            self._advance()  # <
            error_end = self.byte_pos
            # Skip rest of line for recovery
            while self.pos < len(self.source) and self._peek() != "\n":
                self._advance()
            raise ParseError("unexpected token", Span(start, error_end))

        # Bare scalar
        return self._read_bare_scalar(start, had_whitespace, had_newline)

    def _read_quoted_string(self, start: int, had_whitespace: bool, had_newline: bool) -> Token:
        """Read a quoted string."""
        self._advance()  # opening "
        text = ""

        while self.pos < len(self.source):
            ch = self._peek()
            if ch == '"':
                self._advance()
                return Token(
                    TokenType.QUOTED, text, Span(start, self.byte_pos), had_whitespace, had_newline
                )
            if ch == "\\":
                escape_start = self.byte_pos
                self._advance()
                escaped = self._advance()
                match escaped:
                    case "n":
                        text += "\n"
                    case "r":
                        text += "\r"
                    case "t":
                        text += "\t"
                    case "\\":
                        text += "\\"
                    case '"':
                        text += '"'
                    case "u":
                        text += self._read_unicode_escape()
                    case _:
                        raise ParseError(
                            f"invalid escape sequence: \\{escaped}",
                            Span(escape_start, self.byte_pos),
                        )
            elif ch in ("\n", "\r"):
                # Unterminated string - include the newline in the span
                self._advance()
                if ch == "\r" and self._peek() == "\n":
                    self._advance()
                raise ParseError("unexpected token", Span(start, self.byte_pos))
            else:
                text += self._advance()

        # EOF without closing quote - error
        raise ParseError("unexpected token", Span(start, self.byte_pos))

    def _read_unicode_escape(self) -> str:
        """Read a unicode escape sequence."""
        if self._peek() == "{":
            self._advance()
            hex_str = ""
            while self._peek() != "}" and self.pos < len(self.source):
                hex_str += self._advance()
            self._advance()
            return chr(int(hex_str, 16))
        else:
            hex_str = ""
            for _ in range(4):
                hex_str += self._advance()
            return chr(int(hex_str, 16))

    def _read_raw_string(self, start: int, had_whitespace: bool, had_newline: bool) -> Token:
        """Read a raw string."""
        self._advance()  # r
        hashes = 0
        while self._peek() == "#":
            self._advance()
            hashes += 1
        self._advance()  # opening "

        text = ""
        close_pattern = '"' + "#" * hashes

        while self.pos < len(self.source):
            if self.source[self.pos : self.pos + len(close_pattern)] == close_pattern:
                for _ in range(len(close_pattern)):
                    self._advance()
                return Token(
                    TokenType.RAW, text, Span(start, self.byte_pos), had_whitespace, had_newline
                )
            text += self._advance()

        raise ParseError("unclosed raw string", Span(start, self.byte_pos))

    def _read_heredoc(self, start: int, had_whitespace: bool, had_newline: bool) -> Token:
        """Read a heredoc."""
        self._advance()  # <
        self._advance()  # <

        delimiter = ""
        while self.pos < len(self.source) and self._peek() != "\n":
            delimiter += self._advance()
        if self.pos < len(self.source):
            self._advance()  # newline

        # Track content start (after the opening line)
        content_start = self.byte_pos

        text = ""
        bare_delimiter = delimiter.split(",")[0]

        while self.pos < len(self.source):
            line = ""
            while self.pos < len(self.source) and self._peek() != "\n":
                line += self._advance()

            # Check for exact match (no indentation)
            if line == bare_delimiter:
                return Token(
                    TokenType.HEREDOC, text, Span(start, self.byte_pos), had_whitespace, had_newline
                )

            # Check for indented closing delimiter
            stripped = line.lstrip(" \t")
            if stripped == bare_delimiter:
                indent_len = len(line) - len(stripped)
                # Dedent the content by stripping up to indent_len from each line
                result = self._dedent_heredoc(text, indent_len)
                return Token(
                    TokenType.HEREDOC,
                    result,
                    Span(start, self.byte_pos),
                    had_whitespace,
                    had_newline,
                )

            text += line
            if self.pos < len(self.source) and self._peek() == "\n":
                self._advance()
                text += "\n"

        # EOF without closing delimiter - error points at the unmatched content
        raise ParseError("unexpected token", Span(content_start, self.byte_pos))

    def _dedent_heredoc(self, content: str, indent_len: int) -> str:
        """Strip up to indent_len whitespace characters from the start of each line."""
        lines = content.split("\n")
        result = []
        for line in lines:
            stripped = 0
            for ch in line:
                if stripped >= indent_len:
                    break
                if ch in (" ", "\t"):
                    stripped += 1
                else:
                    break
            result.append(line[stripped:])
        return "\n".join(result)

    def _read_bare_scalar(self, start: int, had_whitespace: bool, had_newline: bool) -> Token:
        """Read a bare scalar."""
        text = ""
        while self.pos < len(self.source):
            ch = self._peek()
            if ch in SPECIAL_CHARS:
                break
            text += self._advance()
        return Token(
            TokenType.SCALAR, text, Span(start, self.byte_pos), had_whitespace, had_newline
        )
